File: frontend/src/api/integrations.js
import client from './client';

// Accounts the user has connected on other platforms, and publishing through
// them. One module per concern, not per platform: Reddit is the only one live,
// and Twitter would be more functions here rather than a second client.

// GET /integrations/reddit/status
// `configured` is the server's side (does this deployment have Reddit
// credentials at all); `connected` is the user's. The UI needs both to tell
// "we cannot do this" apart from "you have not connected yet".
export async function getRedditStatus() {
  const { data } = await client.get('/integrations/reddit/status');
  return data;
}

// POST /integrations/reddit/authorize — the URL to send the user to.
// The app never sees or asks for their Reddit password.
export async function getRedditAuthorizeUrl() {
  const { data } = await client.post('/integrations/reddit/authorize');
  return data.authorize_url;
}

// DELETE /integrations/reddit — also revokes the token on Reddit's side.
export async function disconnectReddit() {
  await client.delete('/integrations/reddit');
}

// POST /posts/{id}/publish/reddit
// The title and body are sent rather than re-derived on the server, so what
// the user read on screen is exactly what lands on the subreddit.
export async function publishPostToReddit(postId, { title, body }) {
  const { data } = await client.post(`/posts/${postId}/publish/reddit`, { title, body });
  return data;
}

File: backend/schemas/integrations.py
from datetime import datetime

from pydantic import BaseModel, Field


class ConnectedAccount(BaseModel):
    """What the app may say about a connected account. Never the token."""

    provider: str
    account_handle: str | None = None
    connected_at: datetime | None = None
    scopes: list[str] = Field(default_factory=list)


class RedditStatusResponse(BaseModel):
    # configured: whether this deployment has Reddit credentials at all. The
    # UI needs to tell "we cannot do this" apart from "you have not connected".
    configured: bool
    connected: bool
    subreddit: str
    account: ConnectedAccount | None = None


class AuthorizeUrlResponse(BaseModel):
    authorize_url: str


class PublishToRedditRequest(BaseModel):
    # The headline makes the Reddit title; the post body is the selftext. Both
    # are sent by the client so what the user saw is what gets posted, rather
    # than the server re-deriving it and posting something slightly different.
    title: str = Field(min_length=1, max_length=500)
    body: str = Field(default="", max_length=40000)


class PublicationResponse(BaseModel):
    platform: str
    url: str
    posted_at: datetime
    subreddit: str | None = None
